go/vt/vtgate/evalengine/fn_time.go

```diff
@@ -1200,6 +1205,54 @@ func (call *builtinMonthName) compile(c *compiler) (ctype, error) {
 	return ctype{Type: sqltypes.VarChar, Col: col, Flag: arg.Flag | flagNullable}, nil
 }
 
+func lastDay(loc *time.Location, dt datetime.DateTime) datetime.Date {
+	ts := dt.Date.ToStdTime(loc)
+	firstDayOfMonth := time.Date(ts.Year(), ts.Month(), 1, 0, 0, 0, 0, loc)
+	lastDayOfMonth := firstDayOfMonth.AddDate(0, 1, -1)
+
+	date := datetime.NewDateFromStd(lastDayOfMonth)
+	return date
+}
+
+func (b *builtinLastDay) eval(env *ExpressionEnv) (eval, error) {
+	date, err := b.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	if date == nil {
+		return nil, nil
+	}
+	dt := evalToDateTime(date, -1, env.now, env.sqlmode.AllowZeroDate())
+	if dt == nil || dt.isZero() {
+		return nil, nil
+	}
+
+	d := lastDay(env.currentTimezone(), dt.dt)
+	if d.IsZero() {
```

Comment:
Can this case happen? We already check above that it's not `nil` or a zero date, so how could it be a zero date here? I think we can remove this check (and the equivalent one from the compiler)?

Reply:
hmm... true. This can not happen. Removed from here.
But I didn't add a check for zero datetime before calling lastDay in compiler_asm. And, there's no check in `lastDay` func for checking that, so added a check before calling `lastDay` and removed the check after that, as it was unnecessary.